core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.create.CreateIndexClusterStateUpdateRequest;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateObserver;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateResponse;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.node.NodeClosedException;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.function.BiFunction;
+
+/**
+ * This class provides primitives for waiting for a configured number of shards
+ * to become active before sending a response on an {@link ActionListener}.
+ */
+public class ActiveShardsObserver extends AbstractComponent {
+
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    public ActiveShardsObserver(final Settings settings, final ClusterService clusterService, final ThreadPool threadPool) {
+        super(settings);
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Creates an {@link ActionListener} that responds to a cluster state update upon index creation and
+     * waits on the specified number of active shards to be started before sending the action's response
+     * over its listener.
+     *
+     * @param request the index creation request
+     * @param actionListener the main listener that is listening for responses from the index creation event
+     * @param onResult a function that takes the cluster state update acknowledged flag and timed out flag as parameters
+     *                 and returns a response to be sent on the listener
+     * @return ActionListener for responding to index creation cluster state events and sending the action response
+     *         over the main listener, after waiting for the requested number of shards to be active
+     */
+    public <T> ActionListener<ClusterStateUpdateResponse> waitForActiveShards(final CreateIndexClusterStateUpdateRequest request,
+                                                                              final ActionListener<T> actionListener,
+                                                                              final BiFunction<Boolean, Boolean, T> onResult) {
+        return new ActionListener<ClusterStateUpdateResponse>() {
+            @Override
+            public void onResponse(ClusterStateUpdateResponse response) {
+                if (response.isAcknowledged()) {
+                    final String indexName = request.index();
```

Comment:
make these parmaters to this method? then it doesn't really need to know about index creation requests.